### ISSUE TITLE
How VSS2DBC signal gets data of other signals of same frame

### ISSUE THREAD

**Issue by Athvaith2828** [eclipse-kuksa/kuksa-can-provider#23]
I have exploring kuksa can providers, amazing work.
I have a doubt on VSS2DBC gets the value of other signals of the same frame,
I understood that a single VSS signal will have DBC2VSS(current) and VSS2DBC(target), 
if i set target value of a VSS signal, 
Does it takes **other signa**l value in the following order ?
1.) check **current_value of the signal**
2.)check **target_valur of the signal** 
3.) check for default values in the json

I have a doubt whether it takes target value or current value
please correct me, if i am wrong

**Comment by erikbosch** [comment]
You are right on the general mapping as described in https://github.com/eclipse-kuksa/kuksa-can-provider/blob/main/mapping/README.md

- dbc2vss: The signal from CAN is injected into Databroker as actual value
- vss2dbc: The target value from Databroker is added to the CAN signal

This is (currently) based on an assumption that actual values only flow in the direction from CAN to Databroker, and target values in direction from Databroker to CAN. We do not have as of today any mechanism to specify that sensor values should be used on CAN or that CAN will provide target/actuator values to Databroker.

This is how values in vss2dbc is handled:

At startup the CAN provider subscribes to target values of all VSS signals listed in `vss2dbc` part of the mapping config

https://github.com/eclipse-kuksa/kuksa-can-provider/blob/973fe305f470db216ceddb619c11df7a77adb7a2/dbcfeederlib/databrokerclientwrapper.py#L202

When CAN provider receives a target value update callback from Databroker it:

* Checks which CAN frames that are involved and should be sent
* Populates the CAN frame

https://github.com/eclipse-kuksa/kuksa-can-provider/blob/973fe305f470db216ceddb619c11df7a77adb7a2/dbcfeederlib/dbc2vssmapper.py#L520

Population use two methods:

- It use default values as base, for our example DBC we have some default values in https://github.com/eclipse-kuksa/kuksa-can-provider/blob/973fe305f470db216ceddb619c11df7a77adb7a2/dbc_default_values.json
- Secondly it takes any other value previously received in the subscription from Databroker, as stored in `last_dbc_value`.

So what will be included in the CAN frame is either default values from JSON or actuator/target values from other `vss2dbc` mapping configurations.

**Comment by Athvaith2828** [comment]
Thanks for the quick reply and answer @erikbosch,

Is there an option, that it will look for current value of the VSS and sends that ?

While testing I have noticed 2 things,

1.) For VSS signals of sensor, Can't able to set the target value but in https://github.com/eclipse-kuksa/kuksa-can-provider/blob/973fe305f470db216ceddb619c11df7a77adb7a2/mapping/vss_4.1/dbc_overlay.vspec#L300C3-L300C7 for type of sensor vss2dbc is defined

2.)I have been using flutter/dart to subscribe to kuksa databroker using kuksa/val/v1 proto rpc subscribe 
this is my subscriber entry

var request = SubscribeRequest();
var entry = SubscribeEntry();
entry.path = 'Vehicle.speed';
entry.fields.add(Field.FIELD_PATH);
entry.fields.add(Field.FIELD_VALUE);
request.entries.add(entry);

Vehicle.speed is actuator and when i subscribe using the fields (field_path , field_value) , I am hoping it sends only current_value (if we want target then the field will be autuator_target)  **but I got value when I update both current and target **.

Is this expected behavior ?

**Comment by erikbosch** [comment]
It would technically not be a problem to extend the syntax of the mapping with something like `valuetype: current` or `valuetype: target` and adapt the implementation accordingly so it subscribes to or publish the "right" value.

For `Vehicle.Powertrain.ElectricMotor.Temperature` I would say that it definitively is an error in the *.vspec/*.json, as we cannot subscribe to or at least not will get any target value updates for actuators.
